pkg/operator/ceph/object/notification/controller.go

```diff
@@ -0,0 +1,240 @@
+		controllerName,
+	)
+	if !isReadyToReconcile {
+		// This handles the case where the Ceph Cluster is gone and we want to delete that CR
+		if !cephOBC.GetDeletionTimestamp().IsZero() && !cephClusterExists {
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, nil
+		}
+		logger.Debug("Ceph cluster not yet present.")
+		return reconcileResponse, nil
+	}
+	r.clusterSpec = &cephCluster.Spec
+
+	// DELETE: the CR was deleted
+	if !cephOBC.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("ObjectBucketClaim %q was deleted", cephOBC.Name)
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	// Populate clusterInfo during each reconcile
+	r.clusterInfo, _, _, err = mon.LoadClusterInfo(r.context, cephCluster.Namespace)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to populate cluster info")
+	}
+
+	// create the notification if needed
+	if err = r.updateCephBucketNotifications(cephOBC); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to create bucket notifications")
+	}
+
+	// Other label based reconciliations should come here
+
+	// Return and do not requeue
+	logger.Debug("OBC done reconciling labels")
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileOBCLabels) updateCephBucketNotifications(cephOBC *bktv1alpha1.ObjectBucketClaim) error {
+	// step 1: get the ObjectBucket and StorageClass
+	namespace := cephOBC.Namespace
+	cephOB := bktv1alpha1.ObjectBucket{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.ObjectBucketName}, &cephOB); err != nil {
+		return errors.Wrapf(err, "failed to retrieve ObjectBucket %q", cephOBC.Spec.ObjectBucketName)
+	}
+	cephSC := storagev1.StorageClass{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.StorageClassName}, &cephSC); err != nil {
+		return errors.Wrapf(err, "failed to retrieve StorageClass %q", cephOBC.Spec.StorageClassName)
+	}
+
+	// step 2: delete all existing notifications
+	p := Provisioner{
+		Client:      r.client,
+		Context:     r.context,
+		ClusterInfo: r.clusterInfo,
+		ClusterSpec: r.clusterSpec,
+	}
+	if err := p.CreateSession(cephOB.Spec.AdditionalState["cephUser"], cephSC.Parameters["objectStoreName"]); err != nil {
+		return errors.Wrapf(err, "failed to create session for bucket notification. bucket: %q", cephOB.Name)
+	}
+	if !p.HasSession() {
+		return errors.Errorf("failed to create session for bucket notification. bucket: %q", cephOB.Name)
+	}
+	err := p.DeleteAll(&cephOB)
+	if err != nil {
+		return errors.Wrapf(err, "failed delete all bucket notification from bucket: %q", cephOB.Name)
+	}
+
+	// Step 3: looking for notifications in the labels
+	for key, name := range cephOBC.Labels {
+		notifyLabels := strings.SplitAfterN(key, notificationLabelPrefix, 2)
+		if len(notifyLabels) > 1 && notifyLabels[1] != "" && name == notifyLabels[1] {
```

Comment:
the notification label is not valid, atleast we should log than ignoring the value?

Reply:
it is valid - the user can add labels for different reasons.
it is just not used for bucket notifications.

Reply:
What I meant was if the label is for bucket notification, but the user has given it wrongly. Please consider the following labels :
bucket-notification-my-notification: ""
bucket-notification-my-notification: blah

Should give error/warning for the user for above cases ?

Reply:
agree. logging a warning and ignoring the label in case of mismatch